File: labs/architecture-examples/angularjs_saltarelle/cs/controllers/TodoCtrl.cs
using System.Collections.Generic;
using ng;

namespace todo
{
    public class TodoCtrl
    {
        private List<TodoItem> todos;
        private readonly TodoScope scope;
        private readonly ITodoStorage todoStorage;
        private IFilterService filterFilter;
        private ILogService log;

        public TodoCtrl(TodoScope scope, ILocationService location, ITodoStorage todoStorage,IFilterService filterFilter, ILogService log)
        {
            this.scope = scope;
            this.todoStorage = todoStorage;
            this.filterFilter = filterFilter;
            this.log = log;

            todos = scope.todos = todoStorage.todos;

            scope.newTodo = "";
            scope.editedTodo = null;

            scope.addTodo = addTodo;
            scope.editTodo = editTodo;
            scope.doneEditing = doneEditing;
            scope.removeTodo = removeTodo;
            scope.clearDoneTodos = clearDoneTodos;
            scope.markAll = markAll;

            scope.watch("todos", onTodos, true);
            scope.watch("location.path()", onPath);

            if (location.path() == "") location.path("/");
            scope.location = location;
            log.log("TodoCtrl created");
        }

        private void onPath(object p)
        {
            log.log("TodoCtrl onPath");
            var path = p as string;
            scope.statusFilter = (path == "/active")
                                     ? new StatusFilter {completed = false}
                                     : (path == "/completed")
                                           ? new StatusFilter {completed = true}
                                           : null;
        }

        private void onTodos()
        {
            log.log("TodoCtrl onTodos");
            //log.log(filterFilter);
            //scope.remainingCount = filterFilter(this.todos, new StatusFilter { completed = false }).length;
            scope.doneCount = todos.Count - scope.remainingCount;
            scope.allChecked = scope.remainingCount > 0;
            todoStorage.todos = todos;
        }

        private void addTodo()
        {
            log.log("TodoCtrl addTodo");
            if (scope.newTodo.Length == 0)
            {
                return;
            }

            todos.Add(new TodoItem
                {
                    title = scope.newTodo,
                    completed = false
                });

            scope.newTodo = "";
        }

        private void editTodo(TodoItem todo)
        {
            log.log("TodoCtrl editTodo");
            scope.editedTodo = todo;
        }

        private void doneEditing(TodoItem todo)
        {
            log.log("TodoCtrl doneEditing");
            scope.editedTodo = null;
            if (string.IsNullOrEmpty(todo.title))
            {
                scope.removeTodo(todo);
            }
        }

        private void removeTodo(TodoItem todo)
        {
            log.log("TodoCtrl removeTodo");
            todos.Remove(todo);
        }

        private void clearDoneTodos()
        {
            log.log("TodoCtrl clearDoneTodos");
            scope.todos = todos = todos.Filter((val) => { return !val.completed; });
        }

        private void markAll(bool done)
        {
            log.log("TodoCtrl markAll");
            foreach (TodoItem todo in todos)
            {
                todo.completed = done;
            }
        }
    }
}

File: labs/architecture-examples/angularjs_typescript/js/controllers/TodoCtrl.ts
/// <reference path='../libs/angular-1.0.d.ts' />
/// <reference path='../services/TodoStorage.ts' />
/// <reference path='../models/TodoItem.ts' />
/// <reference path='../interfaces/ITodoStorage.ts' />
/// <reference path='../interfaces/ITodoScope.ts' />

'use strict';

/**
 * The main controller for the app. The controller:
 * - retrieves and persist the model via the todoStorage service
 * - exposes the model to the template and provides event handlers
 */
class TodoCtrl {
    private todos;

    constructor(private $scope: ITodoScope, $location: ng.ILocationService, private todoStorage: ITodoStorage, private filterFilter) {
        this.todos = $scope.todos = todoStorage.get();

        $scope.newTodo = "";
        $scope.editedTodo = null;

        $scope.addTodo = () => this.addTodo();
        $scope.editTodo = (t) => this.editTodo(t);
        $scope.doneEditing = (t) => this.doneEditing(t);
        $scope.removeTodo = (t) => this.removeTodo(t);
        $scope.clearDoneTodos = () => this.clearDoneTodos();
        $scope.markAll = (d) => this.markAll(d);


        $scope.$watch('todos', () => this.onTodos(), true);
        $scope.$watch('location.path()', (path) => this.onPath(path));

        if ($location.path() === '') $location.path('/');
        $scope.location = $location;
    }

    onPath(path) {
        this.$scope.statusFilter = (path == '/active') ?
          { completed: false } : (path == '/completed') ?
            { completed: true } : null;
    }

    onTodos() {
        this.$scope.remainingCount = this.filterFilter(this.todos, { completed: false }).length;
        this.$scope.doneCount = this.todos.length - this.$scope.remainingCount;
        this.$scope.allChecked = !this.$scope.remainingCount
        this.todoStorage.put(this.todos);
    }

    addTodo() {
        if (!this.$scope.newTodo.length) {
            return;
        }

        this.todos.push({
            title: this.$scope.newTodo,
            completed: false
        });

        this.$scope.newTodo = '';
    };

    editTodo(todo: TodoItem) {
        this.$scope.editedTodo = todo;
    };

    doneEditing(todo: TodoItem) {
        this.$scope.editedTodo = null;
        if (!todo.title) {
            this.$scope.removeTodo(todo);
        }
    };

    removeTodo(todo: TodoItem) {
        this.todos.splice(this.todos.indexOf(todo), 1);
    };

    clearDoneTodos() {
        this.$scope.todos = this.todos = this.todos.filter((val) => {
            return !val.completed;
        });
    };

    markAll(done: bool) {
        this.todos.forEach((todo: TodoItem) => {
            todo.completed = done;
        });
    };
}

